README.md
# jsonvalidate

<!-- badges: start -->
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![Build Status](https://travis-ci.org/ropensci/jsonvalidate.svg?branch=master)](https://travis-ci.org/ropensci/jsonvalidate)
[![codecov.io](https://codecov.io/github/ropensci/jsonvalidate/coverage.svg?branch=master)](https://codecov.io/github/ropensci/jsonvalidate?branch=master)
[![](http://www.r-pkg.org/badges/version/jsonvalidate)](https://cran.r-project.org/package=jsonvalidate)
<!-- badges: end -->


Validate JSON against a schema using [`is-my-json-valid`](https://github.com/mafintosh/is-my-json-valid) or [`ajv`](https://github.com/ajv-validator/ajv).  This package is a thin wrapper around these node libraries, using the [V8](https://cran.r-project.org/package=V8) package.

## Usage

Directly validate `json` against `schema`

```r
jsonvalidate::json_validate(json, schema)
```

or create a validator for multiple uses

```r
validate <- jsonvalidate::json_validator(schema)
validate(json)
validate(json2) # etc
```

See the [package vignette](https://docs.ropensci.org/jsonvalidate/articles/jsonvalidate.html) for complete examples.

## Installation

Install from CRAN with

```r
install.packages("jsonvalidate")
```

Alternatively, the current development version can be installed from GitHub with

```r
devtools::install_github("ropensci/jsonvalidate")
```

## License

MIT + file LICENSE © [Rich FitzJohn](https://github.com/richfitz).

 Please note that this project is released with a [Contributor Code of Conduct](https://github.com/ropensci/jsonvalidate/blob/master/CODE_OF_CONDUCT.md). By participating in this project you agree to abide by its terms.

[![ropensci_footer](https://ropensci.org//public_images/github_footer.png)](https://ropensci.org/)


NEWS.md
# jsonvalidate 1.4.1

* Add support for subfolders in nested schema references. (#61)

# jsonvalidate 1.4.0

* Support for safely serialising objects to json, guided by the schema, with new function `json_serialise`
* New object `json_schema` for construction of reusable validation and serialisation functions

# jsonvalidate 1.3.2

* Always uses ES5 version of Ajv, which allows use in both current and "legacy" V8 (#51)

# jsonvalidate 1.3.0

* Upgrade to ajv version 8.5.0
* Add arg `strict` to `json_validate` and `json_validator` to allow evaluating schema in strict mode for ajv only. This is off (`FALSE`) by default to use permissive behaviour detailed in JSON schema

# jsonvalidate 1.2.3

* Schemas can use references to other files with JSON pointers i.e. schemas can reference parts of other files e.g. `definitions.json#/definitions/hello`
* JSON can be validated against a subschema (#18, #19, @AliciaSchep)
* Validation with `error = TRUE` now returns `TRUE` (not `NULL)` on success
* Schemas can span multiple files, being included via `"$ref": "filename.json"` - supported with the ajv engine only (#20, #21, @r-ash).
* Validation can be performed against a fraction of the input data (#25)

# jsonvalidate 1.1.0

* Add support for JSON schema draft 06 and 07 using the [`ajv`](https://github.com/ajv-validator/ajv) node library.  This must be used by passing the `engine` argument to `json_validate` and `json_validator` at present (#2, #11, #15, #16, #17, @karawoo & @ijlyttle)

# jsonvalidate 1.0.1

* Initial CRAN release


js/in.js
import "core-js/es/set";
import "core-js/es/map";
import "core-js/features/array/find";
import "core-js/features/array/find-index";
import "core-js/features/array/from";
import "core-js/features/array/includes";
import "core-js/features/object/assign";
import "core-js/features/string/starts-with";
import "core-js/features/string/includes";

global.Ajv = require('ajv');
global.AjvSchema4 = require('ajv-draft-04');
global.AjvSchema6 = require('ajv/dist/refs/json-schema-draft-06.json');
global.AjvSchema2019 = require('ajv/dist/2019')
global.AjvSchema2020 = require('ajv/dist/2020')
global.addFormats = require('ajv-formats');

global.imjv = require('is-my-json-valid');

global.ajv_create_object = function(meta_schema_version, strict) {
    // Need to disable strict mode, otherwise we get warnings
    // about unknown schema entries in draft-04 (e.g., presence of
    // const) and draft-07 (e.g. presence of "reference")
    var opts = {allErrors: true,
                verbose: true,
                unicodeRegExp: false,
                strict: strict,
                code: {es5: true}};
    if (meta_schema_version === "draft-04") {
        // Need to drop keywords present in later schema versions,
        // otherwise they seem to be not ignored (e.g., a schema that
        // has the 'const' keyword will check it, even though that
        // keyword is not part of draft-04)
        var ret = new AjvSchema4(opts)
            .removeKeyword('propertyNames')
            .removeKeyword('contains')
            .removeKeyword('const')
            .removeKeyword('if')
            .removeKeyword('then')
            .removeKeyword('else');
    } else if (meta_schema_version === "draft/2019-09") {
        var ret = new AjvSchema2019(opts);
    } else if (meta_schema_version === "draft/2020-12") {
        var ret = new AjvSchema2020(opts);
    } else {
        var ret = new Ajv(opts);
        if (meta_schema_version === "draft-06") {
            ret.addMetaSchema(AjvSchema6);
        }
    }
    addFormats(ret);
    return ret;
}

// TODO: we can push greedy into here
global.ajv_create = function(meta_schema_version, strict, schema,
                             filename, dependencies, reference) {
    var ret = ajv_create_object(meta_schema_version, strict);

    if (dependencies) {
        dependencies.forEach(
            function(x) {
                // Avoid adding a dependency and then adding it again as the
                // main schema. This might occur if we have recusive references.
                if (x.id !== filename) {
                    ret.addSchema(drop_id(x.value), x.id)
                }
            });
    }

    if (reference === null) {
        ret = ret.addSchema(drop_id(schema), filename).getSchema(filename);
    } else {
        ret = ret.addSchema(drop_id(schema), filename).getSchema(reference);
    }

    // Save in the global scope so we can use this later from R
    global.validator = ret;
}

global.drop_id = function(x) {
    delete x.id;
    delete x.$id;
    return x;
}

global.imjv_create = function(meta_schema_version, schema) {
    // https://github.com/mafintosh/is-my-json-valid/issues/160
    if (meta_schema_version != "draft-04") {
        throw new Error("Only draft-04 json schema is supported");
    }
    global.validator = imjv(schema);
}

global.ajv_call = function(value, errors, query) {
    var success = validator(jsonpath_eval(value, query));
    var errors = (!success && errors ? validator.errors : null);
    return {"success": success, "errors": errors, "engine": "ajv"};
}

global.imjv_call = function(value, errors, greedy) {
    var success = validator(value, {"greedy": greedy}, {"verbose": errors});
    var errors = (!success && errors ? validator.errors : null);
    return {"success": success, "errors": errors, "engine": "imjv"};
}

global.get_meta_schema_version = function(schema) {
    return schema.$schema;
};

global.find_reference = function(x) {
    var deps = [];

    var f = function(x) {
        if (Array.isArray(x)) {
            // need to descend into arrays as they're used for things
            // like oneOf or anyOf constructs.
            x.forEach(f);
        } else if (typeof(x) === "object" && x !== null) {
            // From the JSON schema docs:
            //
            // > You will always use $ref as the only key in an
            // > object: any other keys you put there will be ignored
            // > by the validator
            //
            // though this turns not to be true empirically...
            if ("$ref" in x) {
                deps.push(x["$ref"]);
            }
            // Would be nicer with Object.values but that does not
            // work on travis apparently.
            Object.keys(x).forEach(function(k) {f(x[k]);});
        }
    };
    f(x);
    return deps;
}

// It might be nice to do this with jsonpath, but that does not seem
// to work well with browserify.  For now, we're going to accept
// 'query' as a string corresponding to a single element
global.jsonpath_eval = function(data, query) {
    if (query === null) {
        return(data);
    }
    if (data === null || Array.isArray(data) || typeof(data) !== "object") {
        throw new Error("Query only supported with object json");
    } else if (!(query in data)) {
        throw new Error("Query did not match any element in the data");
    }
    return data[query];
}


global.typeIsAtomic = function(t) {
    // the const one might be overly generous; it might be that a
    // constant non-atomic type is allowed.
    return t === "string" || t === "number" || t === "boolean" ||
        t == "enum" || t == "const" || t == "integer";
}

global.unboxable = function(x) {
    return Array.isArray(x) && x.length === 1 && typeIsAtomic(typeof(x[0]));
}

global.fixUnboxable = function(x, schema) {
    var f = function(x, s, parent, index) {
        if (Array.isArray(x)) {
            if (typeIsAtomic(s.type) && unboxable(x)) {
                if (parent === null) {
                    x = x[0];
                } else {
                    parent[index] = x[0];
                }
            } else {
                var descend = x.length > 0 &&
                    typeof(x[0]) == "object" &&
                    s.hasOwnProperty("items") &&
                    (s.items.type === "object" || s.items.type == "array");
                if (descend) {
                    for (var i = 0; i < x.length; ++i) {
                        f(x[i], s.items, x, i);
                    }
                }
            }
        } else if (typeof(x) === "object" && x !== null) {
            if (s.type === "object") {
                var keys = Object.keys(s.properties);
                for (var i = 0; i < keys.length; ++i) {
                    var k = keys[i];
                    if (x.hasOwnProperty(k)) {
                        f(x[k], s.properties[k], x, k);
                    }
                }
            }
        }
        return x;
    }

    x = f(x, schema.schema, null);
    return x;
}

global.safeSerialise = function(x) {
    var x = JSON.parse(x);
    return JSON.stringify(fixUnboxable(x, validator));
}
